torch/utils/data/_utils/index_utils.py

```diff
@@ -0,0 +1,150 @@
+from math import sqrt, floor
+
+
+class Range:
+    size: int
+
+    # for slicing
+    start: int
+    stop: int
+    step: int
+
+    # for iterating
+    index: int
+
+    def __init__(self, size: int, start: int = None, stop: int = None, step: int = None):
+        self.size = size
+        self.start = start if start is not None else 0
+        self.stop = stop if stop is not None else size
+        self.step = step if step is not None else 1
+        if not (0 <= self.start <= self.stop):
+            raise ValueError('not (0 <= start <= stop)')
+        if self.stop > self.size:
+            pass  # Allow wrap-around
+        if self.step <= 0:
+            raise ValueError('step <= 0')
+        if (size == 0) != (len(self) == 0):
+            raise ValueError('if size==0 then stop-start must also be 0')
+
+    def __len__(self):
+        if self.start == self.stop:
+            return 0
+        return (self.stop - self.start - 1) // self.step + 1
+
+    def __getitem__(self, index):
+        if isinstance(index, slice):
+            sl = slice(index.start if index.start is not None else 0,
+                       index.stop if index.stop is not None else len(self),
+                       index.step if index.step is not None else 1)
+            if not (0 <= sl.start and sl.start <= sl.stop and 0 < sl.step):
+                raise IndexError
+            return self._slice(sl)
+
+        if not 0 <= index < len(self):
+            raise IndexError
+
+        abs_index = (self.start + index * self.step) % self.size
+
+        return self._get(abs_index)
+
+    def __iter__(self):
+        self.index = 0
+        return self
```

Comment:
It's better to return a iterator instance rather than self. If two iterator called at same time, index race can happen.